examples_tests/50.RadixSort/SortAndScatter.comp

```diff
@@ -0,0 +1,111 @@
+#version 450
+
+#define _NBL_GLSL_WORKGROUP_SIZE_ (1 << 8)
+
+layout (local_size_x = _NBL_GLSL_WORKGROUP_SIZE_) in;
+
+layout (set = 0, binding = 0, std430) readonly buffer in_buffer
+{
+	uvec2 in_values[];
+};
+
+layout (set = 0, binding = 1, std430) writeonly buffer out_buffer
+{
+	uvec2 out_values[];
+};
+
+layout (set = 0, binding = 2, std430) readonly buffer histogram_buffer
+{
+	uint histogram[];
+};
+
+layout(push_constant) uniform pushConstants
+{
+    layout (offset = 0) uint shift;
+} u_push_constants;
+
+#include <nbl/builtin/glsl/workgroup/arithmetic.glsl>
+
+#define BITS_PER_PASS 4
+#define NUM_BUCKETS (1 << BITS_PER_PASS)
+
+#define scratch_shared _NBL_GLSL_SCRATCH_SHARED_DEFINED_
+#define WG_DIM _NBL_GLSL_WORKGROUP_SIZE_
+
+shared uvec2 pingpong[WG_DIM * 2]; // 4kB
+shared uint local_histogram[NUM_BUCKETS]; // 64 bytes
+shared uint global_histogram[NUM_BUCKETS]; // 64 bytes
+
+void main()
+{	
+	if (gl_LocalInvocationID.x < NUM_BUCKETS)
+		global_histogram[gl_LocalInvocationID.x] = histogram[gl_LocalInvocationID.x * gl_NumWorkGroups.x + gl_WorkGroupID.x];
+
+	pingpong[gl_LocalInvocationID.x] = in_values[gl_GlobalInvocationID.x];
+
+	barrier();
+
+	// Locally sort the workgroup with 1bit per pass parallel radix sort
+	// 
+	// Note: Theres an optimization on this with 2 bits per pass and packed keys, as described in the 
+	// following paper. This probably will be useful when we decide to do muliple values per thread.
+	// http://www.heterogeneouscompute.org/wordpress/wp-content/uploads/2011/06/RadixSort.pdf
+	
+	uint offset = 0;
+	
+	for (uint i = 0; i < BITS_PER_PASS; ++i)
+	{
+		uint key = pingpong[offset * WG_DIM + gl_LocalInvocationID.x].x;
+		uint value = pingpong[offset * WG_DIM + gl_LocalInvocationID.x].y;
+		uint shift = u_push_constants.shift + i;
+
+		uint bit = (key >> shift) & 1;
+		uint predicate = 1 - bit; // (bit == 0) ? 1 : 0;
+	
+		scratch_shared[gl_LocalInvocationID.x] = nbl_glsl_workgroupInclusiveAdd(predicate);
+
+		barrier();
+
+		uint total = scratch_shared[WG_DIM - 1];
+		uint prefix_sum = (gl_LocalInvocationID.x == 0) ? 0 : scratch_shared[gl_LocalInvocationID.x - 1];
+
+		uint scatter_address = (bit == 0) ? prefix_sum : (gl_LocalInvocationID.x - prefix_sum + total);
+		
+		offset = 1 - offset;
+		
+		pingpong[offset * WG_DIM + scatter_address] = uvec2(key, value);
+		
+		barrier();
+	}
+
+	uint key = pingpong[WG_DIM * offset + gl_LocalInvocationID.x].x;
+	uint value = pingpong[WG_DIM * offset + gl_LocalInvocationID.x].y;
+	uint digit = (key >> u_push_constants.shift) & 0xf;
+
+	if (gl_LocalInvocationID.x < NUM_BUCKETS)
+		local_histogram[gl_LocalInvocationID.x] = 0;
+
+	barrier();
+
+	// Todo: Is there a better way to generate histogram here, than shared memory atomics?
+
+	atomicAdd(local_histogram[digit], 1);
+
+	barrier();
```

Comment:
you've already computed the local prefix sums/per-bucket-histograms, remember?

**they're the output of the first stage of this whole algorithm!**

Now if you merged this compute kernel with the last downweep (global prefix sum propagation) pass, then you wouldn't need to compute the local histogram again!

Reply:
my bad you cant merge the last downsweep.

Reply:
in this case you want to do this local intra-workgroup sort in the FIRST dispatch, not the last (because you'll still have the local histogram hanging around).

```glsl
uvec2 data = inout_values[gl_GlobalInvocationIndex];
const uint radix = (data.x>>radixOffset)&radixMask;

uint offset;
uint local_prefix_sum[NUM_BUCKETS];
for (int i = 0; i < NUM_BUCKETS; i++)
{
   const bool thisRadix = radix==i;
   nbl_glsl_workgroupBallot(thisRadix);
   local_prefix_sum[i] = nbl_glsl_workgroupBallotInclusiveBitCount();
   if (thisRadix) // something nice for the driver, so it doesnt think we're dynamically indexing local variables
      offset = local_prefix_sum[i];
}

if (gl_LocalInvocationIndex==_NBL_GLSL_WORKGROUP_SIZE_-1u)
{
   // fun fact, because the values will be locally sorted, if you were to prefix sum them again, you'd only care about the reductions, not local prefix sums
   for (int i = 1; i < NUM_BUCKETS; i++)
      histogram[i][gl_WorkgroupIndex] = local_prefix_sum[i];

   // this is a sort of weird in-place last-upsweep
   uint sum = 0;
   scratch_shared[0] = sum;
   for (int i = 1; i < NUM_BUCKETS; i++)
      scratch_shared[i] = sum+=local_prefix_sum[i-1];
}
barrier();
// now you can get a exclusive sum across buckets major, and local invocations minor, a weird in-place downsweep
const uint lastElement = scratch_shared[radix]; // needed to get `lastElement` first because we have overlapping shared memory
//barrier(); // but there's an implict barrier at the start at `nbl_glsl_workgroupShuffle`
offset = nbl_glsl_workgroupShuffle(offset,(gl_LocalInvocationIndex-1)&(_NBL_GLSL_WORKGROUP_SIZE_-1u));
if (gl_LocalInvocationIndex!=0u)
   offset += lastElement;
   
// local sort: cannot use `nbl_glsl_workgroupShuffle` for this because its a gather, and we need a scatter
for (int i=0; i<2; i++)
{
   scratch_shared[offset] = data[i];
   barrier();
   data = scratch_shared[gl_LocalInvocationIndex];
}
// beautifully coalesced store into global memory
inout_values[gl_GlobalInvocationIndex] = data;
```

then your last scatter shader (after all the upsweeps and downsweeps) would do this
```glsl
if (gl_LocalInvocationID.x < NUM_BUCKETS)
		global_histogram[gl_LocalInvocationID.x] = histogram[gl_LocalInvocationID.x * gl_NumWorkGroups.x + gl_WorkGroupID.x];
barrier();

/*
Load your data, compute the radix, etc
*/

uint reconstructedLocalPrefix;
for (int i = 0; i < NUM_BUCKETS; i++)
{
   const bool thisRadix = radix==i;
   nbl_glsl_workgroupBallot(thisRadix);
   uint tmp = nbl_glsl_workgroupBallotExclusiveBitCount(); // use exclusive this time
   if (thisRadix)
      reconstructedLocalPrefix = tmp;
}
const uint scatter_address = global_histogram[radix]+reconstructedLocalPrefix;
out_values[scatter_address] = data;
```

Reply:
yeah I remembered that, but for me the trade off was shared memory atomics for recomputation vs storing the local histograms in global memory in the first shader and then reading them back again in the last shader, and I chose the former

i'm working on the optimization you suggested, seems pretty good till now

Reply:
the advantage of doing a local WG sort in the very first pass, is that it cuts the memory required for storing the `histogram` by literally the size of a workgroup